x/qgb/orchestrator/orchestrator.go

```diff
@@ -0,0 +1,198 @@
+package orchestrator
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+
+	paytypes "github.com/celestiaorg/celestia-app/x/payment/types"
+	"github.com/celestiaorg/celestia-app/x/qgb/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	ethcmn "github.com/ethereum/go-ethereum/common"
+	rpctypes "github.com/tendermint/tendermint/rpc/core/types"
+	coretypes "github.com/tendermint/tendermint/types"
+)
+
+type orchestrator struct {
+	*client
+
+	// celestia related signing
+	signer *paytypes.KeyringSigner
+}
+
+func (oc *orchestrator) orchestrateValsets(ctx context.Context) error {
+	results, err := oc.tendermintRPC.Subscribe(ctx, "valset-changes", "tm.event='Tx' AND message.module='qgb'")
+	if err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case ev := <-results:
+			err = oc.processValsetEvents(ctx, ev)
+			if err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func (oc *orchestrator) processValsetEvents(ctx context.Context, ev rpctypes.ResultEvent) error {
+	attributes := ev.Events[types.EventTypeValsetRequest]
+	for _, attr := range attributes {
+		if attr != types.AttributeKeyNonce {
+			continue
+		}
+
+		queryClient := types.NewQueryClient(oc.qgbRPC)
+
+		lastValsetResp, err := queryClient.LastValsetRequests(ctx, &types.QueryLastValsetRequestsRequest{})
+		if err != nil {
+			return err
+		}
+
+		// todo: double check that the first validator set is found
+		if len(lastValsetResp.Valsets) < 1 {
+			return errors.New("no validator sets found")
+		}
+
+		valset := lastValsetResp.Valsets[0]
+
+		signBytes, err := valset.SignBytes(oc.bridgeID)
+		if err != nil {
+			return err
+		}
+
+		signature, err := oc.personalSignerFn(oc.evmAddress, signBytes.Bytes())
+		if err != nil {
+			return err
+		}
+
+		// create and send the valset hash
+		msg := &types.MsgValsetConfirm{
+			Orchestrator: oc.signer.GetSignerInfo().GetAddress().String(),
+			EthAddress:   oc.evmAddress.Hex(),
+			Nonce:        valset.Nonce,
+			Signature:    ethcmn.Bytes2Hex(signature),
+		}
+
+		err = oc.broadcastTx(ctx, msg)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (oc *orchestrator) orchestrateDataCommitments(ctx context.Context) error {
+	queryClient := types.NewQueryClient(oc.qgbRPC)
+
+	resp, err := queryClient.Params(ctx, &types.QueryParamsRequest{})
+	if err != nil {
+		return err
+	}
+
+	params := resp.Params
+
+	results, err := oc.tendermintRPC.Subscribe(ctx, "height", coretypes.EventQueryNewBlockHeader.String())
+	if err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case ev := <-results:
+			err = oc.processDataCommitmentEvents(ctx, params.DataCommitmentWindow, ev)
+			if err != nil {
+				oc.logger.Err(err)
+			}
+		}
+	}
+}
+
+func (oc *orchestrator) processDataCommitmentEvents(ctx context.Context, window uint64, ev rpctypes.ResultEvent) error {
```

Comment:
I guess it makes sense to change the name since we have a similar method for relayer and might lead to confusion (even if no exported).

Reply:
personally I think having one function called `orchestratrateDataCommitments` which calls another function called `orchestratrateDataCommitmentsEvents` even more confusing :shrug: In both the orchestrator and the relayer, they are processing those types of events. Ideally, we would have an interface for this and everything would be abstracted out cleanly, but I didn't think that was worth the time for the mvp.

its a toss up either way, so happy to rename if you still think otherwise